pkg/terraform/resource/sensitive.go

```diff
@@ -0,0 +1,200 @@
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	errCannotExpandWildcards = "cannot expand wildcards"
+
+	errFmtCannotGetStringForFieldPath  = "cannot not get a string for fieldpath %q"
+	errFmtCannotGetStringsForFieldPath = "cannot get strings matching field fieldpath: %q"
+)
+
+// GetConnectionDetails returns strings matching provided field paths in the
+// input data.
+// See the unit tests for examples.
+func GetConnectionDetails(from map[string]interface{}, at map[string]string) (map[string][]byte, error) {
+	vals := make(map[string][]byte)
+	for tf := range at {
+		vs, err := stringsMatchingFieldPath(from, tf)
+		if err != nil {
+			return nil, errors.Wrapf(err, errFmtCannotGetStringsForFieldPath, tf)
+		}
+		for k, v := range vs {
+			vals[k] = v
+		}
+	}
+	return vals, nil
+}
+
+func stringsMatchingFieldPath(from map[string]interface{}, path string) (map[string][]byte, error) {
+	paved := fieldpath.Pave(from)
+	segments, err := paved.ExpandWildcards(path)
+	if err != nil {
+		return nil, errors.Wrap(err, errCannotExpandWildcards)
+	}
+
+	res := make(map[string][]byte, len(segments))
+	for _, s := range segments {
+		v, err := paved.GetString(s)
+		if err != nil {
+			return nil, errors.Wrapf(err, errFmtCannotGetStringForFieldPath, s)
+		}
+		res[s] = []byte(v)
+	}
+	return res, nil
+}
+
+// GetSensitiveParameters will collect sensitive information as terraform state
+// attributes by following secret references in the spec.
+func GetSensitiveParameters(ctx context.Context, client SecretClient, from runtime.Object, into map[string]interface{}, mapping map[string]string) error {
+	pv, err := fieldpath.PaveObject(from)
+	if err != nil {
+		return err
+	}
+	xpParams := map[string]interface{}{}
+	if err = pv.GetValueInto("spec.forProvider", &xpParams); err != nil {
+		return err
+	}
+	paveXP := fieldpath.Pave(xpParams)
+	paveTF := fieldpath.Pave(into)
+
+	var sensitive []byte
+	for tf, xp := range mapping {
+		expXPs, err := paveXP.ExpandWildcards(xp)
+		if err != nil {
+			return errors.Wrapf(err, "cannot expand wildcard for xp resource")
+		}
+		for _, expXP := range expXPs {
+			sel := v1.SecretKeySelector{}
+			if err = paveXP.GetValueInto(expXP, &sel); err != nil {
+				return errors.Wrapf(err, "cannot get SecretKeySelector from xp resource for fieldpath %q", expXP)
+			}
+			sensitive, err = client.GetSecretValue(ctx, sel)
+			if err != nil {
+				return err
+			}
+			expTF, err := expandedTFPath(expXP, mapping)
+			if err != nil {
+				return err
+			}
+			if err = paveTF.SetString(expTF, string(sensitive)); err != nil {
+				return errors.Wrapf(err, "cannot set string as terraform attribute for fieldpath %q", tf)
+			}
+		}
+	}
+
+	return nil
+}
+
+// GetSensitiveObservation will return sensitive information as terraform state
+// attributes by reading them from connection details.
+func GetSensitiveObservation(from runtime.Object, into map[string]interface{}, mapping map[string]string) error {
```

Comment:
I need to fix this, sensitive observation should come from conn details secret not status.atProvider.

Reply:
Fixed!